src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,388 @@
+            }
+        }
+        return false;
+    }
+
+    private boolean hasCallAsChild(Node n) {
+        for (BackEdges.Edge edge : BackEdges.getOuts(n)) {
+            Node node = edge.node;
+            if (node.getOpCode() == binding_irnode.ir_opcode.iro_Call) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private int dominates(Pointer source, Pointer target) {
+        int dominates = 0;
+
+        if (!dominateMap.containsKey(source)) {
+            // we have no entry yet, compute both values
+            dominates = binding_irdom.block_dominates(source, target);
+
+            HashMap<Pointer, Integer> cachedResults = new HashMap<>();
+            cachedResults.put(target, dominates);
+            dominateMap.put(source, cachedResults);
+
+        } else {
+            Map<Pointer, Integer> cachedResults = dominateMap.get(source);
+            dominates = cachedResults.computeIfAbsent(target, p -> binding_irdom.block_dominates(source, target));
+        }
+        return dominates;
+    }
+
+    private boolean dominatesUses(Node newNode, Node orig) {
+        for (BackEdges.Edge edge : BackEdges.getOuts(orig)) {
+            Node n = edge.node;
+            if (dominates(newNode.getBlock().ptr, n.getBlock().ptr) == 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+
+    private final class CSEVisitor extends NodeVisitor.Default {
+
+        private final Map<TargetValue, Node> constCache = new HashMap<>();
+        private final Map<Entity, Node> addressCache = new HashMap<>();
+        private final Map<NodePreds, Node> nodeCache = new HashMap<>();
+
+        @RequiredArgsConstructor
+        private class NodePreds {
+            private final Node[] preds;
+            private final binding_irnode.ir_opcode opcode;
+            private final Mode mode;
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                NodePreds nodePreds = (NodePreds) o;
+                return Arrays.equals(preds, nodePreds.preds) && opcode == nodePreds.opcode && mode.equals(nodePreds.mode);
+            }
+
+            @Override
+            public int hashCode() {
+                int result = Objects.hash(opcode, mode);
+                result = 31 * result + Arrays.hashCode(preds);
+                return result;
+            }
+        }
+
+        @Override
+        public void defaultVisit(Node node) {
+            // change nothing
+        }
+
+        private void visitPreds(Node node, Node[] preds, binding_irnode.ir_opcode opcode) {
+            NodePreds np = new NodePreds(preds, opcode, node.getMode());
+            if (nodeCache.containsKey(np)) {
+                Node n = nodeCache.get(np);
+                // can exchange node with already found in list
+                nodeValues.put(node, n);
```

Comment:
It might be a good idea to check which of `node` / `n` dominates the other (if any) and put the dominating one in the cache.